File: backend/app/services/include_chunk.py
from __future__ import annotations

import re
from dataclasses import dataclass

# Per-turn include slice. Never the whole vault.
INCLUDE_TURN_CHAR_CAP = 10_000
INCLUDE_TURN_CHAR_MAX = 12_000

INCLUDE_MODES = ("auto", "selection", "heading", "chunk")

_MD_HEADING = re.compile(r"(?m)^(#{1,6})\s+(.+?)\s*$")
_HTML_HEADING = re.compile(r"(?is)<h([1-6])\b[^>]*>(.*?)</h\1>")
_TAG = re.compile(r"(?s)<[^>]+>")


@dataclass(frozen=True)
class IncludeSlice:
    text: str
    label: str
    chars: int
    mode: str
    offset: int
    next_offset: int | None
    next_heading: str | None
    has_more: bool

    @property
    def chip(self) -> str:
        return format_include_chip(self.label, self.chars)


def format_include_chip(label: str, chars: int) -> str:
    title = (label or "Slice").strip() or "Slice"
    return f"§ {title} ({chars:,} chars)"


def _strip_html(raw: str) -> str:
    text = _HTML_HEADING.sub(
        lambda match: "\n" + ("#" * int(match.group(1))) + " " + _TAG.sub("", match.group(2)).strip() + "\n",
        raw or "",
    )
    text = _TAG.sub(" ", text)
    return re.sub(r"[ \t]+\n", "\n", re.sub(r"\n{3,}", "\n\n", text)).strip()


def normalize_include_body(raw: str | None, html: str | None = None) -> str:
    text = (raw or "").strip()
    if _MD_HEADING.search(text):
        return text
    if html and _HTML_HEADING.search(html):
        return _strip_html(html)
    if _HTML_HEADING.search(text):
        return _strip_html(text)
    return text


def parse_sections(body: str) -> list[tuple[str, int, int]]:
    """Return (title, start, end) ranges in `body`, including a lead-in Opening if needed."""
    text = body or ""
    matches = list(_MD_HEADING.finditer(text))
    if not matches:
        return [("Opening", 0, len(text))] if text else []
    sections: list[tuple[str, int, int]] = []
    if matches[0].start() > 0 and text[: matches[0].start()].strip():
        sections.append(("Opening", 0, matches[0].start()))
    for index, match in enumerate(matches):
        end = matches[index + 1].start() if index + 1 < len(matches) else len(text)
        title = match.group(2).strip() or "Section"
        sections.append((title, match.start(), end))
    return sections


def clamp_slice(text: str, cap: int = INCLUDE_TURN_CHAR_CAP) -> str:
    cleaned = text or ""
    limit = min(max(1, cap), INCLUDE_TURN_CHAR_MAX)
    if len(cleaned) <= limit:
        return cleaned
    cut = cleaned[:limit]
    if " " in cut:
        cut = cut.rsplit(" ", 1)[0]
    return cut.rstrip()


def _needs_chunk(body: str) -> bool:
    return len(body or "") > INCLUDE_TURN_CHAR_CAP


def resolve_include_slice(
    body: str,
    *,
    mode: str | None = None,
    selection: str | None = None,
    heading: str | None = None,
    offset: int = 0,
    title: str | None = None,
    html: str | None = None,
) -> IncludeSlice:
    source = normalize_include_body(body, html)
    cleaned_mode = (mode or "auto").strip().lower()
    if cleaned_mode not in INCLUDE_MODES:
        cleaned_mode = "auto"
    fallback_label = (title or "").strip() or "Included"

    if cleaned_mode == "selection":
        quote = re.sub(r"\s+", " ", (selection or "").strip())
        if not quote:
            raise ValueError("Highlight text in the reader, then Include selection.")
        text = clamp_slice(quote)
        found = source.find(quote[: min(len(quote), 80)]) if source else -1
        start = found if found >= 0 else 0
        end = start + len(text)
        has_more = end < len(source)
        next_heading = None
        if has_more:
            for label, sec_start, _sec_end in parse_sections(source):
                if sec_start >= end:
                    next_heading = label
                    break
        return IncludeSlice(
            text=text,
            label="Selection",
            chars=len(text),
            mode="selection",
            offset=start,
            next_offset=end if has_more else None,
            next_heading=next_heading,
            has_more=has_more,
        )

    sections = parse_sections(source)

    if cleaned_mode == "heading":
        want = (heading or "").strip().lower()
        if not want:
            raise ValueError("Pick a heading from this note.")
        match = next((item for item in sections if item[0].strip().lower() == want), None)
        if match is None:
            match = next((item for item in sections if want in item[0].strip().lower()), None)
        if match is None:
            raise ValueError(f"No heading named {heading!r} in this note.")
        label, start, end = match
        chunk = clamp_slice(source[start:end])
        next_heading = None
        next_offset = None
        has_more = False
        for other_label, other_start, _other_end in sections:
            if other_start >= end:
                next_heading = other_label
                next_offset = other_start
                has_more = True
                break
        if not has_more and start + len(chunk) < len(source):
            has_more = True
            next_offset = start + len(chunk)
        return IncludeSlice(
            text=chunk,
            label=label,
            chars=len(chunk),
            mode="heading",
            offset=start,
            next_offset=next_offset,
            next_heading=next_heading,
            has_more=has_more,
        )

    start = max(0, int(offset or 0))
    if start >= len(source) and source:
        start = 0
    remainder = source[start:]
    chunk = clamp_slice(remainder)
    end = start + len(chunk)
    has_more = end < len(source)
    label = fallback_label
    next_heading = None
    for sec_label, sec_start, sec_end in sections:
        if sec_start <= start < sec_end or (start == 0 and sec_start == 0):
            if sec_label != "Opening":
                label = sec_label
        if has_more and sec_start >= end and next_heading is None:
            next_heading = sec_label
    if start == 0 and not _needs_chunk(source):
        label = fallback_label
        has_more = False
        next_heading = None
    return IncludeSlice(
        text=chunk,
        label=label,
        chars=len(chunk),
        mode="chunk" if has_more or start else ("chunk" if _needs_chunk(source) else "auto"),
        offset=start,
        next_offset=end if has_more else None,
        next_heading=next_heading,
        has_more=has_more,
    )


def format_excerpt(title: str, slice: IncludeSlice) -> str:
    header = f"Title: {(title or 'Untitled').strip()}\nSlice: {slice.chip}"
    if slice.has_more:
        header += "\nHas more: yes — Steve can ask for the next chunk."
    return f"{header}\n\n{slice.text}"


def slice_meta(slice: IncludeSlice) -> dict[str, object]:
    payload: dict[str, object] = {
        "include_chip": slice.chip,
        "include_label": slice.label,
        "include_chars": slice.chars,
        "include_mode": slice.mode,
        "include_offset": slice.offset,
        "include_has_more": slice.has_more,
    }
    if slice.next_offset is not None:
        payload["include_next_offset"] = slice.next_offset
    if slice.next_heading:
        payload["include_next_heading"] = slice.next_heading
    return payload

File: frontend/src/lib/include-chunk.ts
export const INCLUDE_TURN_CHAR_CAP = 10_000;
export const INCLUDE_TURN_CHAR_MAX = 12_000;
export const WORKING_NOTE_CHAR_CAP = 80_000;

export type IncludeMode = "auto" | "selection" | "heading" | "chunk";

export type ArticleSection = {
  title: string;
  start: number;
  end: number;
};

export type IncludeSlice = {
  text: string;
  label: string;
  chars: number;
  mode: IncludeMode;
  offset: number;
  nextOffset: number | null;
  nextHeading: string | null;
  hasMore: boolean;
  chip: string;
};

const MD_HEADING = /^(#{1,6})\s+(.+?)\s*$/gm;

export function formatIncludeChip(label: string, chars: number): string {
  const title = (label || "Slice").trim() || "Slice";
  return `§ ${title} (${chars.toLocaleString("en-US")} chars)`;
}

export function parseSections(body: string): ArticleSection[] {
  const text = body || "";
  const matches = [...text.matchAll(MD_HEADING)];
  if (!matches.length) {
    return text ? [{ title: "Opening", start: 0, end: text.length }] : [];
  }
  const sections: ArticleSection[] = [];
  const first = matches[0]!;
  if (first.index && first.index > 0 && text.slice(0, first.index).trim()) {
    sections.push({ title: "Opening", start: 0, end: first.index });
  }
  matches.forEach((match, index) => {
    const start = match.index ?? 0;
    const end = index + 1 < matches.length ? (matches[index + 1]!.index ?? text.length) : text.length;
    sections.push({ title: (match[2] || "Section").trim(), start, end });
  });
  return sections;
}

export function articleNeedsIncludeSlice(body: string | null | undefined, cap = INCLUDE_TURN_CHAR_CAP): boolean {
  return (body || "").length > cap;
}

export function clampIncludeSlice(text: string, cap = INCLUDE_TURN_CHAR_CAP, hardMax = INCLUDE_TURN_CHAR_MAX): string {
  const limit = Math.min(Math.max(1, cap), hardMax);
  if (text.length <= limit) return text;
  let cut = text.slice(0, limit);
  const space = cut.lastIndexOf(" ");
  if (space > limit * 0.6) cut = cut.slice(0, space);
  return cut.trimEnd();
}

export function resolveIncludeSlice(input: {
  body: string;
  mode?: IncludeMode | null;
  selection?: string | null;
  heading?: string | null;
  offset?: number;
  title?: string | null;
  cap?: number;
  hardMax?: number;
}): IncludeSlice {
  const source = input.body || "";
  const mode = input.mode || "auto";
  const fallback = (input.title || "").trim() || "Included";
  const turnCap = input.cap ?? INCLUDE_TURN_CHAR_CAP;
  const ceiling = input.hardMax ?? (input.cap ?? INCLUDE_TURN_CHAR_MAX);
  const clip = (text: string) => clampIncludeSlice(text, turnCap, ceiling);

  if (mode === "selection") {
    const quote = (input.selection || "").replace(/\s+/g, " ").trim();
    const text = clip(quote);
    const found = quote ? source.indexOf(quote.slice(0, Math.min(quote.length, 80))) : -1;
    const start = found >= 0 ? found : 0;
    const end = start + text.length;
    const hasMore = end < source.length;
    let nextHeading: string | null = null;
    if (hasMore) {
      nextHeading = parseSections(source).find((section) => section.start >= end)?.title ?? null;
    }
    return {
      text,
      label: "Selection",
      chars: text.length,
      mode: "selection",
      offset: start,
      nextOffset: hasMore ? end : null,
      nextHeading,
      hasMore,
      chip: formatIncludeChip("Selection", text.length),
    };
  }

  const sections = parseSections(source);
  if (mode === "heading") {
    const want = (input.heading || "").trim().toLowerCase();
    const match =
      sections.find((section) => section.title.toLowerCase() === want) ||
      sections.find((section) => section.title.toLowerCase().includes(want));
    const start = match?.start ?? 0;
    const end = match?.end ?? Math.min(source.length, start + turnCap);
    const text = clip(source.slice(start, end));
    const next = sections.find((section) => section.start >= end);
    const hasMore = Boolean(next) || start + text.length < source.length;
    const label = match?.title || fallback;
    return {
      text,
      label,
      chars: text.length,
      mode: "heading",
      offset: start,
      nextOffset: next?.start ?? (hasMore ? start + text.length : null),
      nextHeading: next?.title ?? null,
      hasMore,
      chip: formatIncludeChip(label, text.length),
    };
  }

  const start = Math.max(0, input.offset || 0);
  const chunk = clip(source.slice(start));
  const end = start + chunk.length;
  const hasMore = end < source.length;
  let label = fallback;
  let nextHeading: string | null = null;
  for (const section of sections) {
    if (section.start <= start && start < section.end && section.title !== "Opening") {
      label = section.title;
    }
    if (hasMore && section.start >= end && !nextHeading) nextHeading = section.title;
  }
  if (start === 0 && source.length <= turnCap) {
    return {
      text: chunk,
      label: fallback,
      chars: chunk.length,
      mode: "auto",
      offset: 0,
      nextOffset: null,
      nextHeading: null,
      hasMore: false,
      chip: formatIncludeChip(fallback, chunk.length),
    };
  }
  return {
    text: chunk,
    label,
    chars: chunk.length,
    mode: "chunk",
    offset: start,
    nextOffset: hasMore ? end : null,
    nextHeading,
    hasMore,
    chip: formatIncludeChip(label, chunk.length),
  };
}

export function readerIncludeContext(root?: HTMLElement | null): { selection: string; heading: string | null } {
  if (typeof window === "undefined") return { selection: "", heading: null };
  const selection = window.getSelection();
  const quote = selection && !selection.isCollapsed ? selection.toString().replace(/\s+/g, " ").trim() : "";
  let heading: string | null = null;
  const articleRoot =
    root ||
    (selection?.rangeCount
      ? (selection.getRangeAt(0).commonAncestorContainer as HTMLElement).closest?.(".article-body")
      : null) ||
    (document.querySelector(".article-body") as HTMLElement | null);
  if (selection && selection.rangeCount && articleRoot) {
    let node: Node | null = selection.getRangeAt(0).startContainer;
    while (node && node !== articleRoot) {
      if (node instanceof HTMLElement && /^H[1-6]$/.test(node.tagName)) {
        heading = node.innerText.replace(/\s+/g, " ").trim();
        break;
      }
      node = node.parentNode;
    }
    if (!heading) {
      const range = selection.getRangeAt(0);
      const walker = document.createTreeWalker(articleRoot, NodeFilter.SHOW_ELEMENT);
      let lastHeading: string | null = null;
      let current: Node | null = walker.currentNode;
      while (current) {
        if (current instanceof HTMLElement && /^H[1-6]$/.test(current.tagName)) {
          lastHeading = current.innerText.replace(/\s+/g, " ").trim();
        }
        if (current === range.startContainer || current.contains(range.startContainer)) break;
        current = walker.nextNode();
      }
      heading = lastHeading;
    }
  }
  return { selection: quote, heading };
}
